-1

**Failed Tests**:   Build  HeaderConsistency  ClangBuild 
**Size**: This PR adds an extra 64KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-193333/42269/summary.html
**COMMIT**: 84e760106cdd1e292cad0bb21aa2d8ead571348a
**CMSSW**: CMSSW_14_2_X_2024-10-16-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46421/42269/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @bsunanda cms-sw/cmssw#46404

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-193333/42269/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-193333/42269/git-merge-result


## Build

I found compilation error when building: 

<pre>>> Compiling  src/FastSimulation/CaloGeometryTools/src/BaseCrystal.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_14_2_X_2024-10-16-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_2_X_2024-10-16-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/cms/cmssw/CMSSW_14_2_X_2024-10-16-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/cms/coral/CORAL_2_3_21-484c499dc51e2646f152cf0149808593/include/LCG -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/dd4hep/v01-29-00-f595e98759e21600901a749c6dd8f4e2/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/boost/1.80.0-477823d53efabc5118f199265eb7ab49/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/clhep/2.4.7.1-17c7283747b6cbc68a3534cb987a890c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/curl/7.79.0-abae79fca02c8e83ab0f428f2a91d96b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/gsl/2.6-4e597de26ced42091883404fa44cef2e/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/lcg/root/6.30.09-c59d6b036f1cbd6988c172ba319259f1/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/tbb/v2021.9.0-e785b749a0b6cb9c66dc1d78066210e0/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/fmt/10.2.1-deba157dd76fed17ac8b42adc60ff53f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/OpenBLAS/0.3.27-6574458f8642b299ac965c654e194a8b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/tinyxml2/6.2.0-68d511689c5dbd9a0d296bbb795fa052/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/FastSimulation/CaloGeometryTools/src/FastSimulationCaloGeometryTools/BaseCrystal.cc.d src/FastSimulation/CaloGeometryTools/src/BaseCrystal.cc -o tmp/el8_amd64_gcc12/src/FastSimulation/CaloGeometryTools/src/FastSimulationCaloGeometryTools/BaseCrystal.cc.o
>> Compiling  src/FastSimulation/CaloGeometryTools/src/CaloDirectionOperations.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_14_2_X_2024-10-16-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_2_X_2024-10-16-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/cms/cmssw/CMSSW_14_2_X_2024-10-16-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/cms/coral/CORAL_2_3_21-484c499dc51e2646f152cf0149808593/include/LCG -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/dd4hep/v01-29-00-f595e98759e21600901a749c6dd8f4e2/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/boost/1.80.0-477823d53efabc5118f199265eb7ab49/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/clhep/2.4.7.1-17c7283747b6cbc68a3534cb987a890c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/curl/7.79.0-abae79fca02c8e83ab0f428f2a91d96b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/gsl/2.6-4e597de26ced42091883404fa44cef2e/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/lcg/root/6.30.09-c59d6b036f1cbd6988c172ba319259f1/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/tbb/v2021.9.0-e785b749a0b6cb9c66dc1d78066210e0/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-39786ff94f8aa2f543922cad720e1b32/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/fmt/10.2.1-deba157dd76fed17ac8b42adc60ff53f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/OpenBLAS/0.3.27-6574458f8642b299ac965c654e194a8b/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02859/el8_amd64_gcc12/external/tinyxml2/6.2.0-68d511689c5dbd9a0d296bbb795fa052/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/FastSimulation/CaloGeometryTools/src/FastSimulationCaloGeometryTools/CaloDirectionOperations.cc.d src/FastSimulation/CaloGeometryTools/src/CaloDirectionOperations.cc -o tmp/el8_amd64_gcc12/src/FastSimulation/CaloGeometryTools/src/FastSimulationCaloGeometryTools/CaloDirectionOperations.cc.o
src/FastSimulation/CaloGeometryTools/src/CaloDirectionOperations.cc: In static member function 'static CaloDirection CaloDirectionOperations::add2d(const CaloDirection&, const CaloDirection&)':
src/FastSimulation/CaloGeometryTools/src/CaloDirectionOperations.cc:6:3: error: 'assert' was not declared in this scope
    6 |   assert(d1 != 999);
      |   ^~~~~~
src/FastSimulation/CaloGeometryTools/src/CaloDirectionOperations.cc:2:1: note: 'assert' is defined in header '<cassert>'; did you forget to '#include <cassert>'?
    1 | #include "FastSimulation/CaloGeometryTools/interface/CaloDirectionOperations.h"
  +++ |+#include <cassert>
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package FWCore/Services
>> Entering Package FastSimulation/CaloGeometryTools
>> Entering Package Geometry/ForwardGeometry
>> Entering Package HLTrigger/JetMET
>> Compile sequence completed for CMSSW CMSSW_14_2_X_2024-10-16-2300
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_2_X_2024-10-16-2300/tmp/el8_amd64_gcc12/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package DataFormats/DTRecHit
Entering library rule at DataFormats/DTRecHit
>> Compiling  src/DataFormats/DTRecHit/src/DTChamberRecSegment2D.cc
</pre>

--- a/DataFormats/DTRecHit/interface/DTDriftTimeParameters.icc
+++ b/DataFormats/DTRecHit/interface/DTDriftTimeParameters.icc
@@ -4,6 +4,7 @@
 // THIS_CLASS should be #defined accordingly.
 
 #include <string>
+#include <cassert>
 
 #define DT_Cell_Width 42
 #define DT_Cell_HalfWidth 21
@@ -473,6 +474,10 @@ inline void THIS_CLASS::MB_DT_Get_grid_values(
       break;
     }
   }
+
+  // To silence LLVM analyzer warning
+  assert(jValue >= 0);
+  assert(iValue >= 0);
 
   if (fabs(Values[iValue] - Var) <= fabs(Var - Values[jValue])) {
     *pi = iValue;

--- a/DataFormats/GeometryCommonDetAlgo/interface/DeepCopyPointerByClone.h
+++ b/DataFormats/GeometryCommonDetAlgo/interface/DeepCopyPointerByClone.h
@@ -2,9 +2,10 @@
 #define DeepCopyPointerByClone_H
 
 #include <algorithm>
+#include <cassert>
 
 /** Same as DeepCopyPointer, except that it copies the object
- *  pointed to wsing the clone() virtual copy constructor.
+ *  pointed to using the clone() virtual copy constructor.
  */
 
 template <class T>
@@ -34,17 +35,29 @@ public:
   }
 
   // straight from http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2006/n2027.html
-  DeepCopyPointerByClone(DeepCopyPointerByClone&& other) : theData(other.theData) { other.theData = 0; }
+  DeepCopyPointerByClone(DeepCopyPointerByClone&& other) : theData(other.theData) { other.theData = nullptr; }
   DeepCopyPointerByClone& operator=(DeepCopyPointerByClone&& other) {
     std::swap(theData, other.theData);
     return *this;
   }
 
-  T& operator*() { return *theData; }
-  const T& operator*() const { return *theData; }
+  T& operator*() {
+    assert(theData);
+    return *theData;
+  }
+  const T& operator*() const {
+    assert(theData);
+    return *theData;
+  }
 
-  T* operator->() { return theData; }
-  const T* operator->() const { return theData; }
+  T* operator->() {
+    assert(theData);
+    return theData;
+  }
+  const T* operator->() const {
+    assert(theData);
+    return theData;
+  }
 
   /// to allow test like " if (p) {...}"
   operator bool() const { return theData != 0; }

--- a/EventFilter/Utilities/plugins/ExceptionGenerator.cc
+++ b/EventFilter/Utilities/plugins/ExceptionGenerator.cc
@@ -178,7 +178,8 @@ namespace evf {
           edm::LogError("TestErrorMessage") << qualifier_;
           break;
         case 8:
-          *pi = 0;  //intentionally caused segfault by assigning null pointer (this produces static-checker warning)
+          [[clang::suppress]] *pi =
+              0;  //intentionally caused segfault by assigning null pointer (this produces static-checker warning)
           break;
         case 9:
           for (unsigned int j = 0; j < intqualifier_ * 1000 * 100; j++) {
@@ -209,8 +210,10 @@ namespace evf {
         case 12: {
           timeval tv_now;
           gettimeofday(&tv_now, nullptr);
-          if ((unsigned)(tv_now.tv_sec - tv_start_.tv_sec) > intqualifier_)
-            *pi = 0;  //intentionally caused segfault by assigning null pointer (this produces static-checker warning)
+          if ((unsigned)(tv_now.tv_sec - tv_start_.tv_sec) > intqualifier_) {
+            [[clang::suppress]] *pi =
+                0;  //intentionally caused segfault by assigning null pointer (this produces static-checker warning)
+          }
         } break;
         case 13: {
           void *vp = malloc(1024);

--- a/FWCore/Framework/src/StreamSchedule.cc
+++ b/FWCore/Framework/src/StreamSchedule.cc
@@ -1189,9 +1189,10 @@ namespace edm {
         ServiceRegistry::Operate operate(weakToken.lock());
 
         if (iPtr) {
-          //this is used to prioritize this error over one
-          // that happens in EndPath or Accumulate
-          pathErrorPtr->store(new std::exception_ptr(*iPtr));
+          // free previous value of pathErrorPtr, if any;
+          // prioritize this error over one that happens in EndPath or Accumulate
+          auto currentPtr = pathErrorPtr->exchange(new std::exception_ptr(*iPtr));
+          assert(currentPtr == nullptr);
         }
         finishedPaths(*pathErrorPtr, std::move(allPathsHolder), transitionInfo);
       });

--- a/FWCore/Services/src/SiteLocalConfigService.cc
+++ b/FWCore/Services/src/SiteLocalConfigService.cc
@@ -185,7 +185,11 @@ namespace edm {
 
     std::filesystem::path const SiteLocalConfigService::storageDescriptionPath(
         edm::CatalogAttributes const &aDataCatalog) const {
-      std::string siteconfig_path = std::string(std::getenv("SITECONFIG_PATH"));
+      char *tmp = std::getenv("SITECONFIG_PATH");
+      if (tmp == nullptr) {
+        throw cms::Exception("SiteLocalConfigService") << "SITECONFIG_PATH is not set!";
+      }
+      std::string siteconfig_path = std::string(tmp);
       std::filesystem::path filename_storage;
       //not a cross site use local path given in SITECONFIG_PATH
       if (aDataCatalog.site == aDataCatalog.storageSite) {

--- a/HLTrigger/JetMET/plugins/HLTRHemisphere.cc
+++ b/HLTrigger/JetMET/plugins/HLTRHemisphere.cc
@@ -177,13 +177,17 @@ void HLTRHemisphere::ComputeHemispheres(std::unique_ptr<std::vector<math::XYZTLo
       if (itemp / j_count == 1) {
         if (count < JETS.size())
           j_temp1 += JETS.at(count);
-        else
+        else {
+          assert(extraJets);  // to silence LLVM analyzer warning
           j_temp1 += extraJets->at(count - JETS.size());
+        }
       } else {
         if (count < JETS.size())
           j_temp2 += JETS.at(count);
-        else
+        else {
+          assert(extraJets);  // to silence LLVM analyzer warning
           j_temp2 += extraJets->at(count - JETS.size());
+        }
       }
       itemp -= j_count * (itemp / j_count);
       j_count /= 2;

